content/slice/language-guide/enum-types.md

```diff
@@ -11,62 +13,121 @@ with an associated integral value. It is similar to enumerations found in langua
+one. In the `Fruit` enum above, `Orange` gets value `6`.
 {% /slice1 %}
 
 {% slice2 %}
+
+## Enumeration type
+
+An enumeration type or enum type is a user-defined type that can take two forms:
+
+### Enum with underlying type
+
+An enum with an underlying type holds a set of named constants called enumerators, and is similar to C++ and C# enums.
+
+The underlying type can be any integral type: both [fixed size][fixed-size] and [variable size][variable-size] integral
+types are accepted. The following example shows how to define an enumeration named `Fruit` with three enumerators:
+`Apple`, `Pear`, and `Orange`. Its underlying type is `uint8`:
+
+```slice
+enum Fruit : uint8 { Apple, Pear, Orange }
+```
+
+### Enum with fields
+
+An enum with no underlying type is called an "enum with fields" because each of its enumerators may define fields. An
+enum with fields is a discriminated union, just like enums in Rust and Swift.
+
+For example:
+
+```slice
+enum Shape { // no underlying type
+    Circle(radius: uint32)
+    Rectangle(width: uint32, length: uint32)
+    Dot // an enumerator without any field
+}
+```
+
+## Enumerators - with underlying type
+
+The enumerators of an enum with an underlying type are integral constants. The default behavior of the Slice compiler is
+to assign values to these enumerators automatically. The first enumerator is assigned a value of `0`, and subsequent
+enumerators are assigned increasing values. The range of these enumerator values is the range of the underlying type.
+
+The following code example illustrates how you can explicitly assign values for the enumerators in your enumeration
+type. In this example, the `Apple` enumerator is assigned a value of `1`, the `Pear` enumerator is assigned a value
+of `5`, and the `Orange` enumerator is left without an explicitly assigned value.
+
 ```slice
 enum Fruit : uint8 {
     Apple = 1
     Pear = 5
     Orange
 }
 ```
-{% /slice2 %}
 
-If you don't assign a value to an enumerator, the value of this enumerator is the value of the previous enumerator plus
-one. In the `Fruit` enum above, `Orange` gets value `6`.
+## Enumerators - with fields
+
+The enumerators of an enum with fields are not constants; you can think of them as instances of nested structs.
+
+And just like for the fields of a struct, you can include tagged fields unless the enclosing enum is marked `compact`.
+For example:
+
+```slice
+// Fields can be tagged unless the enum if marked compact
+enum FlagColor {
+    Red(tag(1) code: uint16?)
+    White
+    Blue(tag(1) code: uint16?)
+}
+
+// A compact enum does not accept tagged fields
+compact enum LaunchResult {
+    Success(speed: float)
+    Failure(message: string, errorCode: int32)
+}
+```
+
+You can also assign a numeric value to each enumerator. These numeric values are called discriminants and are used to
+encode and decode the enumeration.
```

Comment:
Do we decode enumerations or the enumerators? Or both.
```suggestion
encode and decode the enumerator.
```

Reply:
We encode/decode the enumeration.

Talking about the enumerator could be confusing:  is the discriminant part of the enumerator? Arguably it's not since it's a constant, and the enumerator itself may not known this value. The discriminant is about the encoding/decoding of the enumeration.

Reply:
I see your point. The discriminant probably isn't part of the enumerator.